File: daily_challenges/3186.cpp
//Approach-1 - Recursion + Memoization
//T.C : O(nlogn)
//S.C : O(n)
class Solution {
public:
    typedef long long ll;
    int n;                                
    unordered_map<ll,ll> mp;      
    vector<ll> t;           

    long long solve(int i, const vector<ll>& nums) {
        if (i >= n) 
            return 0;

        if (t[i] != -1)
            return t[i];

        //skip current damage
        ll skip = solve(i + 1, nums);

        //take current damage
        int j = lower_bound(begin(nums) + i + 1, end(nums), nums[i] + 3) - begin(nums);
        ll take = nums[i] * mp[nums[i]] + solve(j, nums);

        return t[i] = max(skip, take);
    }

    long long maximumTotalDamage(vector<int>& power) {
        for (int x : power) 
            mp[x]++;

        vector<ll> nums(mp.size());

        for (auto &p : mp)
            nums.push_back(p.first);

        sort(begin(nums), end(nums));
        n = nums.size();
        t.assign(n, -1);

        return solve(0, nums);
    }
};

File: daily_challenges/3186.js


/**
 * @param {number[]} power
 * @return {number}
 */
var maximumTotalDamage = function (power) {
    const mp = new Map()
    for (let p of power) {
        mp.set(p, (mp.get(p) || 0) + 1)
    }
    const nums = Array.from(mp.keys()).sort((a, b) => a - b)
    const n = nums.length
    const t = new Array(n).fill(-1)

    function solve(i) {
        if (i >= n) return 0
        if (t[i] !== -1) return t[i]
        const skip = solve(i + 1)
        const target = nums[i] + 3
        const j = lowerBound(nums, target, i + 1)
        const take = nums[i] * mp.get(nums[i]) + solve(j)
        t[i] = Math.max(take, skip)
        return t[i]
    }
    function lowerBound(arr, target, start) {
        let low = start
        let high = arr.length
        while (low < high) {
            let mid = Math.floor((low + high) / 2)
            if (arr[mid] < target) {
                low = mid + 1
            } else {
                high = mid
            }
        }
        return low
    }
    return solve(0)
};